Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp

```diff
@@ -19,96 +19,366 @@
+			else if (bNextColsSatisfiesMinCells &&
+				(NextRowsRowColRatioDiff >= NextColsRowColRatioDiff || !bNextRowsSatisfiesMinCells))
+			{
+				Cols = NextCols;
+			}
+			else
+			{
+				break;
+			}
+		}
+
+		const float GridWidth = Cols * CellSize;
+		const float GridHeight = Rows * CellSize;
+
+		const float StartX = WorldPositon.X - GridWidth / 2.0f + CellSize / 2.0f;
+		const float StartY = WorldPositon.Y - GridHeight / 2.0f + CellSize / 2.0f;
+
+		OutCellCenterPoints.Empty();
+		for (int32 Row = 0; Row < Rows; ++Row)
+		{
+			for (int32 Col = 0; Col < Cols; ++Col)
+			{
+				const float PointX = Col * CellSize + StartX;
+				const float PointY = Row * CellSize + StartY;
+				OutCellCenterPoints.Add({ PointX, PointY, 0.0f });
+			}
+		}
+
+		return true;
+	}
 } // anonymous namespace
 
-ABenchmarkGymGameMode::ABenchmarkGymGameMode()
-	: bInitializedCustomSpawnParameters(false)
-	, NumPlayerClusters(1)
-	, PlayersSpawned(0)
-	, NPCSToSpawn(0)
+// --- USpawnCluster ---
+
+USpawnCluster::~USpawnCluster()
 {
-	PrimaryActorTick.bCanEverTick = true;
+	for (AActor* SpawnPointActor : SpawnPointActors)
+	{
+		SpawnPointActor->Destroy();
+	}
+}
 
-	static ConstructorHelpers::FClassFinder<AActor> DropCubeClassFinder(TEXT("/Game/Benchmark/DropCube_BP"));
-	DropCubeClass = DropCubeClassFinder.Class;
+void USpawnCluster::GenerateSpawnPoints()
+{
+	const int32 DistBetweenSpawnPoints = 300;
+	const bool bSuccefullyCreatedReducedGrid = GenerateMinimalGridInArea(Width, Height, MaxSpawnPoints, DistBetweenSpawnPoints, WorldPosition, SpawnPoints);
+	if (!bSuccefullyCreatedReducedGrid)
+	{
+		return;
+	}
+
+	// Sort the spawn points by "distance from centre of cluster" ascending.
+	// This ensures we iterate from closest to furthest spawn point.
+	SpawnPoints.Sort([this](const FVector& Point1, const FVector& Point2)
+	{
+		const float Distance1 = FVector::Distance(Point1, WorldPosition);
+		const float Distance2 = FVector::Distance(Point2, WorldPosition);
+		return Distance2 > Distance1;
+	});
+
+	//Remove any excess spawn points
+	SpawnPoints.RemoveAt(MaxSpawnPoints, SpawnPoints.Num() - MaxSpawnPoints);
 }
 
-void ABenchmarkGymGameMode::GenerateTestScenarioLocations()
+const TArray<AActor*>& USpawnCluster::CreateSpawnPointActors()
 {
-	constexpr float RoamRadius = 7500.0f; // Set to half the NetCullDistance
+	UWorld* World = GetWorld();
+
+	for (int i = 0; i < SpawnPoints.Num(); ++i)
 	{
-		FRandomStream PlayerStream;
-		PlayerStream.Initialize(FCrc::MemCrc32(&ExpectedPlayers, sizeof(ExpectedPlayers))); // Ensure we can do deterministic runs
-		for (int i = 0; i < ExpectedPlayers; i++)
-		{
-			FVector PointA = PlayerStream.VRand()*RoamRadius;
-			FVector PointB = PlayerStream.VRand()*RoamRadius;
-			PointA.Z = PointB.Z = 0.0f;
-			PlayerRunPoints.Emplace(FBlackboardValues{ PointA, PointB });
-		}
+		const FVector& SpawnPoint = SpawnPoints[i];
+
+		FActorSpawnParameters SpawnInfo;
+		SpawnInfo.Owner = World->GetAuthGameMode();
+		SpawnInfo.Instigator = NULL;
```

Comment:
Nullptr?

Reply:
For some reason unreal natively initializes it as NULL. Not sure why. I have remove the setting of SpawnInfo.Instigator as it is redundant.

Reply:
Probably legacy engine code from unreal 3 days or even earlier.